client/x/evmstaking/keeper/withdraw_test.go

```diff
@@ -0,0 +1,502 @@
+			},
+			withdraw: &bindings.IPTokenStakingWithdraw{
+				DelegatorCmpPubkey: delPubKey1.Bytes(),
+				ValidatorCmpPubkey: valPubKey.Bytes(),
+				Amount:             new(big.Int).SetUint64(1),
+				Raw:                gethtypes.Log{},
+			},
+		},
+		{
+			name: "fail: invalid delegator pubkey",
+			withdraw: &bindings.IPTokenStakingWithdraw{
+				DelegatorCmpPubkey: delPubKey2.Bytes()[:16],
+				ValidatorCmpPubkey: valPubKey.Bytes(),
+				Amount:             new(big.Int).SetUint64(1),
+				Raw:                gethtypes.Log{},
+			},
+			expectedErr: "invalid pubkey length",
+		},
+		{
+			name: "fail: invalid validator pubkey",
+			withdraw: &bindings.IPTokenStakingWithdraw{
+				DelegatorCmpPubkey: delPubKey2.Bytes(),
+				ValidatorCmpPubkey: valPubKey.Bytes()[:16],
+				Amount:             new(big.Int).SetUint64(1),
+				Raw:                gethtypes.Log{},
+			},
+			expectedErr: "invalid pubkey length",
+		},
+		{
+			name: "fail: corrupted delegator pubkey",
+			withdraw: &bindings.IPTokenStakingWithdraw{
+				DelegatorCmpPubkey: createCorruptedPubKey(delPubKey1.Bytes()),
+				ValidatorCmpPubkey: valPubKey.Bytes(),
+				Amount:             new(big.Int).SetUint64(1),
+				Raw:                gethtypes.Log{},
+			},
+			expectedErr: "delegator pubkey to evm address",
+		},
+		{
+			name: "fail: corrupted validator pubkey",
+			withdraw: &bindings.IPTokenStakingWithdraw{
+				DelegatorCmpPubkey: delPubKey1.Bytes(),
+				ValidatorCmpPubkey: createCorruptedPubKey(valPubKey.Bytes()),
+				Amount:             new(big.Int).SetUint64(1),
+				Raw:                gethtypes.Log{},
+			},
+			expectedErr: "validator pubkey to evm address",
+		},
+		{
+			name: "fail: unknown depositor",
+			settingMock: func() {
+				accountKeeper.EXPECT().HasAccount(gomock.Any(), sdk.AccAddress(unknownPubKey.Address().Bytes())).Return(false).Times(1)
+			},
+			withdraw: &bindings.IPTokenStakingWithdraw{
+				DelegatorCmpPubkey: unknownPubKey.Bytes(),
+				ValidatorCmpPubkey: valPubKey.Bytes(),
+				Amount:             new(big.Int).SetUint64(1),
+				Raw:                gethtypes.Log{},
+			},
+			expectedErr: "depositor account not found",
+		},
+		{
+			name: "fail: amount to withdraw is greater than the delegation amount",
+			settingMock: func() {
+				accountKeeper.EXPECT().HasAccount(gomock.Any(), sdk.AccAddress(delPubKey2.Address().Bytes())).Return(true).Times(1)
+			},
+			withdraw: &bindings.IPTokenStakingWithdraw{
+				DelegatorCmpPubkey: delPubKey2.Bytes(),
+				ValidatorCmpPubkey: valPubKey.Bytes(),
+				Amount:             new(big.Int).SetUint64(math.MaxUint64),
+				Raw:                gethtypes.Log{},
+			},
+			expectedErr: "invalid shares amount",
+		},
+	}
+
+	for _, tc := range tcs {
+		s.Run(tc.name, func() {
+			if tc.settingMock != nil {
+				tc.settingMock()
+			}
+			err := keeper.ProcessWithdraw(ctx, tc.withdraw)
+			if tc.expectedErr != "" {
+				require.ErrorContains(err, tc.expectedErr)
+			} else {
+				require.NoError(err)
```

Comment:
Can we check that [undelegate message](https://github.com/piplabs/story/blob/main/client/x/evmstaking/keeper/withdraw.go#L280) is sent?

Reply:
@limengformal 
That's a good point 🙂 
I added check whether unbonding delegation is successfully created or not. (link. https://github.com/piplabs/story/pull/65/commits/d511d7e5200b41412e9bf6297e44296f2f07a42b)
If it is created, it means undelegate meesage sent and well processed.